src/cpu/aarch64/acl_gemm_convolution.hpp

```diff
@@ -105,13 +107,23 @@ struct acl_gemm_convolution_fwd_t : public primitive_t {
                 "gemm:acl", acl_gemm_convolution_fwd_t, USE_GLOBAL_SCRATCHPAD);
 
         status_t init(engine_t *engine) {
-            bool ok = true && is_fwd()
+            using namespace data_type;
+            using smask_t = primitive_attr_t::skip_mask_t;
+
+            bool ok = is_fwd()
                     && set_default_alg_kind(alg_kind::convolution_direct)
-                    && expect_data_types(data_type::f32, data_type::f32,
-                            data_type::f32, data_type::f32, data_type::f32)
+                    && expect_data_types(
+                            src_type, wei_type, bia_type, dst_type, undef)
                     && !has_zero_dim_memory()
-                    && attr()->has_default_values(
-                            primitive_attr_t::skip_mask_t::post_ops)
+                    && IMPLICATION(with_bias(),
+                            IMPLICATION(
+                                    src_type == u8, bias_md_.data_type == s32))
+                    && IMPLICATION(src_type == f32, bias_md_.data_type == f32)
```

Comment:
I am not sure I understand the need of these implications. As far as I understand, the bias data type is fixed in the implementation (set via instantiation of the template parameter). Which means that `expect_data_types(src_type, ...., bia_type, ...)` will automatically imply the truth of this expression:
- `IMPLICATION(with_bias(), IMPLICATION(src_type == u8, bia_type == s32))`
    - it is worth noting that currently it will always be true just because `src_type` is never `u8`
- `IMPLICATION(with_bias(), IMPLICATION(src_type == f32, bia_type == f32))`

---

It is worth noting the minor difference between generic reference convolution and this ACL one.
In generic reference implementation the template parameters are: `src_type, wei_type, dst_type` and (!) `acc_type`. See [here](https://github.com/oneapi-src/oneDNN/blob/b6d8518d326d95f78e00d3b7ecffd245e490695c/src/cpu/ref_convolution.hpp#L35-L36). And they can work with different bias types using the casting at runtime (see [`get_bias()`](https://github.com/oneapi-src/oneDNN/blob/b6d8518d326d95f78e00d3b7ecffd245e490695c/src/cpu/ref_convolution.cpp#L238)).
Of course not all combinations make sense. That's why there are some checks for bias in [`pd_t::init()`](https://github.com/oneapi-src/oneDNN/blob/b6d8518d326d95f78e00d3b7ecffd245e490695c/src/cpu/ref_convolution.hpp#L54-L59). Notice, that on contrary with this implementation the 3rd parameter of [`expect_data_types()`](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/ref_convolution.hpp#L49) that corresponds to bias is `undef`, meaning this function doesn't check the bias at all.

Your implementation is much more strict. Each instantiation fixes the bias data type it can work with (checked via `expect_data_types()`. So, you either don't need these extra checks (apart from, maybe, whether there is or there is no bias at all), or you might want to remove the bias from the template to reduce the number of implementation instantiations.

Reply:
Thanks for taking the time to explain this, I get it now. I chose to remove the `pd_t::init()` checks and leave the bias in the template for now.